thirtyinch/src/main/java/net/grandcentrix/thirtyinch/internal/PresenterSavior.java

```diff
@@ -15,53 +15,226 @@
+                final String presenterId = entry.getKey();
+                final TiPresenter presenter = entry.getValue();
+
+                // when the presenter is not destroyed yet, destroy it.
+                if (!presenter.isDestroyed()) {
+                    presenter.destroy();
+                }
+                scope.remove(presenterId);
+            }
+        }
+
+        printRemainingPresenter();
     }
 
     @Override
     @Nullable
-    public TiPresenter recover(final String presenterId) {
-        return mPresenters.get(presenterId);
+    public TiPresenter recover(final String presenterId, @NonNull final Activity activity) {
+        final ActivityScopedPresenters scope = getScope(activity);
+        if (scope == null) {
+            return null;
+        }
+        return scope.get(presenterId);
     }
 
     @Override
-    public String save(@NonNull final TiPresenter presenter) {
-        final String id = generateId(presenter);
-        TiLog.v(TAG, "save presenter with id " + id + " " + presenter);
-        mPresenters.put(id, presenter);
+    public String save(@NonNull final TiPresenter presenter, @NonNull final Activity activity) {
+        final String id = generatePresenterId(presenter);
+
+        final ActivityScopedPresenters scope = getScopeOrCreate(activity);
+        scope.save(id, presenter);
+        printRemainingPresenter();
+
         return id;
     }
 
-    private String generateId(@NonNull final TiPresenter presenter) {
+    private String generatePresenterId(@NonNull final TiPresenter presenter) {
         return presenter.getClass().getSimpleName()
                 + ":" + presenter.hashCode()
                 + ":" + System.nanoTime();
     }
+
+    /**
+     * retrieves an existing scope for a {@link Activity} but doesn't create on when the scope
+     * doesn't exist
+     */
+    @Nullable
+    private synchronized ActivityScopedPresenters getScope(final Activity activity) {
+        final ActivityInstanceObserver detector = mActivityInstanceObserver;
+        if (detector == null) {
+            return null;
+        }
+
+        final String scopeId = detector.getActivityId(activity);
+        if (scopeId == null) {
+            return null;
+        }
+
+        return mScopes.get(scopeId);
+    }
+
+    /**
+     * Returns an existing scope or creates a new one for the given Activity. Registers a callback
+     * so the scope will be cleaned up when the Activity finishes
+     *
+     * @return an existing or new created scope for presenters
+     */
+    @NonNull
+    private synchronized ActivityScopedPresenters getScopeOrCreate(final Activity activity) {
+        if (mActivityInstanceObserver == null) {
+            mActivityInstanceObserver = new ActivityInstanceObserver(this);
+            TiLog.v(TAG, "registering lifecycle callback");
+            activity.getApplication().registerActivityLifecycleCallbacks(mActivityInstanceObserver);
+        }
```

Comment:
Please extract this statement to a method like `registerObserverAsLifecyclerCallback`